thewalrus/internal_modes/pnr_statistics.py

```diff
@@ -120,3 +126,121 @@ def pnr_prob(covs, i, hbar=2):
+    for index in np.ndindex(*repeats_p):
+        coeff_pref = 1.0 + 0j
+        for i, block in enumerate(blocks):
+            (coeff, val) = finite_difference_operator_coeffs(repeats_p[i] - 1, index[i])
+            coeff_pref *= coeff
+            for mode in block:
+                coeff_vect[mode] = val
+        AX_S = get_AX_S(coeff_vect, A)
+
+        netsum += coeff_pref * f_from_matrix(AX_S, 2 * n)[n]
+    return netsum
+
+
+def probabilities_single_mode(cov, pattern, normalize=False, LO_overlap=None, cutoff=13, hbar=2):
+    """
+    Calculates the diagonal of the density matrix, hence the name probabilities, of first mode when heralded by pattern on a zero-displaced, M-mode Gaussian state
+    where each mode contains K internal modes.
+
+    Args:
+        cov (array): 2MK x 2MK covariance matrix
+        pattern (dict): heralding pattern total photon number in the spatial modes (int), indexed by spatial mode
+        normalize (bool): whether to normalise the output density matrix
+        LO_overlap (array): overlap between internal modes and local oscillator
+        cutoff (int): photon number cutoff. Should be odd. Even numbers will be rounded up to an odd number
+        hbar (float): the value of hbar (default 2)
+    Returns:
+        array[complex]: (cutoff+1, cutoff+1) dimension density matrix
+    """
+
+    # The lines until A =  Amat(...) are copies from density_single_mode
+    cov = np.array(cov).astype(np.float64)
+    M = len(pattern) + 1
+    K = cov.shape[0] // (2 * M)
+    if not set(list(pattern.keys())).issubset(set(list(np.arange(M)))):
+        raise ValueError("Keys of pattern must correspond to all but one spatial mode")
+    N_nums = np.array(list(pattern.values()))
+    HM = list(set(list(np.arange(M))).difference(list(pattern.keys())))[0]
+    if LO_overlap is not None:
+        if not K == LO_overlap.shape[0]:
+            raise ValueError("Number of overlaps with LO must match number of internal modes")
+        if not (np.linalg.norm(LO_overlap) < 1 or np.allclose(np.linalg.norm(LO_overlap), 1)):
+            raise ValueError("Norm of overlaps must not be greater than 1")
+
+    # swapping the spatial modes around such that we are heralding in spatial mode 0
+    Uswap = np.zeros((M, M))
+    swapV = np.concatenate((np.array([HM]), np.arange(HM), np.arange(HM + 1, M)))
+    for j, k in enumerate(swapV):
+        Uswap[j][k] = 1
+    U_K = np.zeros((M * K, M * K))
+    for i in range(K):
+        U_K[i::K, i::K] = Uswap
+    _, cov = passive_transformation(np.zeros(cov.shape[0]), cov, U_K, hbar=hbar)
+
+    cov = project_onto_local_oscillator(cov, M, LO_overlap=LO_overlap, hbar=hbar)
+
+    A = Amat(cov)
+    Q = Qmat(cov)
+    fact = 1 / np.sqrt(np.linalg.det(Q).real)
+    blocks = np.arange(K * M).reshape([M, K])
+    probs = np.zeros([cutoff])
+    patt = [pattern[i] for i in range(1, 1 + len(pattern))]
+    for i in range(cutoff):
+        patt_long = [i] + patt
```

Comment:
```suggestion
    for i in range(cutoff):
        patt_long = [i] + N_nums
```
At the moment this assumes the PNRs are on all but the first mode, but in the case that the output mode is not the top one, the covariance matrix has been permuted and the pattern has not.